brave/src/main/java/brave/handler/SpanListener.java

```diff
@@ -0,0 +1,106 @@
+ * when a span is created or abandoned. The purpose of this type is to allow tracking of children,
+ * or partitioning of data for backend that needs to see an entire {@linkplain
+ * TraceContext#localRootId() local root}.
+ *
+ * <p>As with {@link FinishedSpanHandler}, it is important to do work quickly as callbacks are run
+ * on the same thread as application code. That said, there are some rules to keep in mind below.
+ *
+ * <p>The {@link TraceContext} parameter from {@link #onCreate} will be the same reference for
+ * all callbacks, except {@link #onOrphan}, which has value, but not reference equality.
+ *
+ * <p>The {@link MutableSpan} parameter from {@link #onCreate} will be the same reference for
+ * all callbacks. Do not mutate {@link MutableSpan} between callbacks as it is not thread safe.
+ *
+ * <p>If caching the {@link TraceContext} parameter, consider a {@link WeakReference} to avoid
+ * holding up garbage collection.
+ */
+public class SpanListener {
+  /** Use to avoid comparing against null references */
+  public static final SpanListener NOOP = new SpanListener() {
+    @Override public String toString() {
+      return "NoopSpanListener{}";
+    }
+  };
+
+  protected SpanListener() {
+  }
+
+  /**
+   * This is called when a span is allocated, but before it is started. An allocation here will
+   * result in one of:
+   *
+   * <ol>
+   *   <li>{@link #onAbandon} if this was a speculative context</li>
+   *   <li>{@link #onFlush} if this was intentionally reported incomplete</li>
+   *   <li>{@link #onOrphan} if this was reported incomplete due to garbage collection</li>
+   *   <li>{@link #onFinish} if this was reported complete</li>
+   * </ol>
+   *
+   * <p>The {@code parent} can be {@code null} only when the new context is a {@linkplain
+   * TraceContext#isLocalRoot() local root}.
+   */
+  public void onCreate(@Nullable TraceContext parent, TraceContext context, MutableSpan span) {
+  }
+
+  /**
+   * Called on {@link Span#abandon()}.
+   *
+   * <p>This is useful when counting children. Decrement your counter when this occurs as the span
+   * will not be reported.
+   *
+   * <p><em>Note:</em>Abandoned spans should be ignored as they aren't indicative of an error. Some
+   * instrumentation speculatively create a span for possible outcomes such as retry.
+   */
+  public void onAbandon(TraceContext context, MutableSpan span) {
+  }
+
+  /**
+   * Called on {@link Span#flush()}.
+   *
+   * <p>Even though the span here will is incomplete (missing {@link MutableSpan#finishTimestamp()},
+   * it is reported to the tracing system unless a {@link FinishedSpanHandler} returns false.
+   */
+  public void onFlush(TraceContext context, MutableSpan span) {
+  }
+
+  /**
+   * Called when the trace context was garbage collected prior to completion.
+   *
+   * <p>Unlike {@link FinishedSpanHandler#supportsOrphans()}, this is called even if {@linkplain
```

Comment:
your questions are helpful as I also feel uncomfortable.

I'm wondering if the type shouldn't be called something else..

We are mostly talking about a binding between a trace context and a state object for the purpose of collection and when that binding is complete (for any reason). When there are
only two states it is simpler and no diagrams are needed.

Plus we don't need to describe nuance (though we can in rationale), such as that spans are sometimes never started or their start timestamp overwritten (ex "start" is not a state in a state diagram, just field manipulation)

Ex
```java
SampledSpanListener {
  enum Reason {
    FINISH,
    FLUSH,
    GC;
  }
  void onBind(TraceContext, MutableSpan)
  void onRelease(TraceContext, MutableSpan, Reason)
}
```

Reply:
This interface does seem nicer since many use cases won't even care about the reason and it's more explicit that only one of those is called. That said, I don't know if conceptually it's so different, docs that would apply to the individual interface methods seem to just move into the Reason enum.

Also unsure of bind, the word isn't very intuitive for me. `onStart` / `onEnd` seem clear, with the caveat of `end` looking a lot like `finish` (but then again `release` I guess looks like `GC` 🤣 ).

Wondering what happens if the methods return `boolean`? `onBind` can return secondary sampling decision, `onRelease` can replace `FinishedSpanHandler` so don't have to reason about that?

Reply:
I think the biggest difference the enum brings is that the state diagram becomes unambiguous.. whatever the begin hook has only one end hook and they are both always called. Those who care about the reason of the end can look at it. I agree docs will just move, but this obviates docs about a state diagram I think.

naming is not my favorite either.. was trying to avoid the word finish. I tried thinking through various things that suggest an association.
ex

associate/disassociate
allocate/deallocate
bind/unbind

random words that just say start/stop
begin/end
start/finish (dangerous as conflated with span ops)

I'm currently thinking..
onSampled
onCollected

wrt to `onCollected` returning false, I think we could I guess yeah, just it would also imply that any finishedSpanListeners would need to be converted first to these in order to perform things like redaction. Another way is we can make this able to internally dispatch to finishedSpanHandlers... ex Whatever this is called exposes a factory that accepts a single (potentially composite) FinishedSpanHandler instance we give to it...

Reply:
one possibility especially with the design as it is, is that we can deprecate FinishedSpanHandler. I feel one thing this design does is makes the relationships a whole lot more clear. So what would happen in this case is that we make a special SampledSpanListener that does a dispatch first to any FinishedSpanHandlers before calling its onCollected hook. that way we don't have to mention `FinishedSpanHandler` at all? It is a little sad that `FinishedSpanHandler` didn't live that long, but I feel this is becoming a much better conceptual model.